libraries/AP_Scripting/AP_Scripting_SerialAccess.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+  generic object to allow a script to use a serial driver stream from both
+  driver and device perspectives
+ */
+
+#include "AP_Scripting_config.h"
+#include "AP_Scripting.h"
+#include "AP_Scripting_SerialAccess.h"
+
+#if AP_SCRIPTING_ENABLED
+
+#if AP_SCRIPTING_SERIALDEVICE_ENABLED
+#define check_is_device_port() (is_device_port)
+#define ON_DEVICE_PORT(func, ...) (((AP_Scripting_SerialDevice::Port*)stream)->device_##func (__VA_ARGS__))
+#else
+#define check_is_device_port() (false)
+#define ON_DEVICE_PORT(...) (0) // not executed
+#endif
+
+void AP_Scripting_SerialAccess::begin(uint32_t baud)
+{
+    if (!check_is_device_port()) {
+        stream->begin(baud);
+    }
+}
+
+size_t AP_Scripting_SerialAccess::write(uint8_t c)
+{
+    return write(&c, 1);
+}
+
+size_t AP_Scripting_SerialAccess::write(const uint8_t *buffer, size_t size)
+{
+    if (!check_is_device_port()) {
+        return stream->write(buffer, size);
+    } else {
+        return ON_DEVICE_PORT(write, buffer, size);
+    }
+}
+
+int16_t AP_Scripting_SerialAccess::read(void)
```

Comment:
Do we *really* need this?  It's *really* awful as an interface!

Reply:
Can you help me understand you are pointing to and what is awful about it?

The ultimate problem is that the objects returned by `serial:find_serial` and `serial:find_simulated_device` must be able to access two different sets of methods on the same C++ object. So either they can each return a different type of userdata (which was done formerly and objected to by @IamPete1) or there needs to be a new class (here `AP_Scripting_SerialAccess`) which chooses dynamically. The latter way does save flash (roughly 300B), so I'd say that one.

Here is the first attempt if you want to compare: https://github.com/tpwrules/ardupilot/tree/ssd-v1

Reply:
Had a brief discussion on the Discord and the objection here is to the `int16_t` return type. It would be less error prone to have something like `bool read(uint8_t&)`. But this would break existing scripts and representing a reference type like this in Lua is not possible (it would have to be a second return value).

In any case I had a plan to revise some other types in a separate PR so I will fold this into that.